Add Xilinx Axi Ethernet Lite

WorldofJARcraft opened the pull request:
This commit adds support for the Xilinx AXI Ethernet Lite device,  also known as the emaclite.
The emaclite is a light-weight 10/100 MII Ethernet device commonly found in Xilinx FPGA / SoC designs, as it does not require a separate license.
It foregoes a DMA to reduce chip area.  Instead, it uses memory-mapped transmit/receive buffers.
A selection of features can optionally be enabled:
- a second ("pong") RX/TX Buffer
- MDIO support
- Interrupt support

This driver handles the MAC functionality of the core; a driver for the MDIO part is introduced separately as an MDIO driver.

Documentation for the device is available [online](https://docs.amd.com/r/en-US/pg135-axi-ethernetlite/Receive-Control-Register-Ping).
This driver has been tested in an out-of-tree variant of the OpenHW Group cva6 SoC (RISC-V 64) on a Digilent Arty A7 board. In particular, the most commonly used version with all optional features enabled has been tested.

Here is an exemplary device tree entry for this device:
```dts
axi-ethernet-lite@40E00000{

		/* shared between MAC and MDIO */
		reg = <0x40E00000 0x10000>;

               compatible = "xlnx,xps-ethernetlite-3.00.a";

		status = "okay";

		arty_ethernet:axi-ethernet-lite-mac {
			/* MAC and MDIO cannot be parent and child, this breaks the preprocessing macros */
			compatible = "xlnx,xps-ethernetlite-3.00.a-mac";
			interrupt-parent = <&plic>;
			interrupts = <12 0x4>;
			clocks = <&clk_bus_0>;
      	
			local-mac-address = [00 0a 35 00 00 01];
			
			xlnx,rx-ping-pong;
			xlnx,tx-ping-pong;
			phy-connection-type = "mii";

			status = "okay";
		};


		axi_ethernetlite_0_mdio: mdio {
			#address-cells = <1>;
			#size-cells = <0>;

			compatible = "xlnx,xps-ethernetlite-3.00.a-mdio";
			status = "okay";

			arty_phy0: phy@1 {
				#address-cells = <1>;
				#size-cells = <0>;
				compatible = "ethernet-phy";
				reg = <1>;
				status="okay";
    			};
	  	};
	};
```

WorldofJARcraft committed:
drivers: mdio: Add Xilinx Axi Ethernet Lite driver

This commit adds support for the Xilinx AXI Ethernet Lite device,
also known as the emaclite.
The emaclite is a light-weight 10/100 MII Ethernet device.
It can optionally be configured to include an MDIO.
The MMIO interface is controlled via MMIO registers and requires
the software to busy-wait until completion.

Signed-off-by: Eric Ackermann <eric.ackermann@cispa.de>

WorldofJARcraft commented:
FYI @venodela

maass-hamburg commented on drivers/ethernet/eth_xilinx_axi_ethernet_lite.c:
```diff
@@ -0,0 +1,613 @@
+	}
+	LOG_INF("MAC address set!");
+
+	if (config->phy) {
```
check with `device_is_ready`

WorldofJARcraft replied:
Done

WorldofJARcraft commented on drivers/ethernet/eth_xilinx_axi_ethernet_lite.c:
```diff
@@ -0,0 +1,613 @@
+	const struct device *phy;
+	const uintptr_t reg;
+
+	/* device tree properties */
+	bool has_duplex;
+	bool has_mdio;
+	bool has_rx_ping_pong;
+	bool has_tx_ping_pong;
+	bool has_random_mac_address;
+	bool has_interrupt;
```
Done

WorldofJARcraft commented on drivers/ethernet/eth_xilinx_axi_ethernet_lite.c:
```diff
@@ -0,0 +1,613 @@
+#define XILINX_OUI_BYTE_1 0x0A
+#define XILINX_OUI_BYTE_2 0x35
+
+static int xilinx_axi_ethernet_lite_probe(const struct device *dev)
```
Done

WorldofJARcraft commented on drivers/mdio/mdio_xilinx_axi_ethernet_lite.c:
```diff
@@ -0,0 +1,223 @@
+
+static DEVICE_API(mdio, mdio_xilinx_axi_ethernet_lite_api) = {
+	/* clause 45 not supported */
+	.bus_enable = xilinx_axi_ethernet_lite_mdio_bus_enable,
+	.bus_disable = xilinx_axi_ethernet_lite_mdio_bus_disable,
```
I call `xilinx_axi_ethernet_lite_mdio_bus_enable` in init now.
I left these functions in, as certain PHYs do call them.

maass-hamburg replied:
These functions are optional, if a phy calls them and they are not implemented, then it won't fail.

WorldofJARcraft replied:
OK, they have been removed.

WorldofJARcraft commented on drivers/mdio/mdio_xilinx_axi_ethernet_lite.c:
```diff
@@ -0,0 +1,223 @@
+#define MDIO_XILINX_AXI_ETHERNET_LITE_MDIO_ADDRESS_REG_SHIFT_REGADDR 0
+#define MDIO_XILINX_AXI_ETHERNET_LITE_MDIO_ADDRESS_REG_SHIFT_PHYADDR 5
+
+struct mdio_xilinx_axi_ethernet_lite_data {
```
Added mutex

WorldofJARcraft commented on drivers/ethernet/eth_xilinx_axi_ethernet_lite.c:
```diff
@@ -0,0 +1,613 @@
+	static const struct xilinx_axi_ethernet_lite_config config_##inst = {                      \
+		.config_func = xilinx_axi_ethernet_lite_config_##inst,                             \
+		.phy = DEVICE_DT_GET_OR_NULL(DT_INST_PHANDLE(inst, phy_handle)),                   \
+		.reg = DT_REG_ADDR(DT_INST_PARENT(inst)),                                          \
```
Done

WorldofJARcraft commented on drivers/mdio/mdio_xilinx_axi_ethernet_lite.c:
```diff
@@ -0,0 +1,223 @@
+		LOG_ERR("MDIO bus not enabled!");
+		return -ENOSYS;
+	}
+	if (mdio_xilinx_axi_ethernet_lite_check_busy(config)) {
+		LOG_ERR("MDIO bus busy!");
+		return -ENOSYS;
```
This serves as an error check: In case the device is completely stuck, there is no point in starting a new transaction. So we throw an error in this case.
If the device is not stuck, the device should never be busy when read/write are called, as they block until the transaction completes.

WorldofJARcraft replied:
Transactions can time out, which releases the mutex.
The busy state can then be encountered by the next - if any - transaction.

Generally, the core is not *supposed to* get stuck in an MDIO transaction. I only expect this to happen, e.g., when it is not correctly wired up (this is commonly used in FPGA designs, so this can happen).
I figured printing a clear error message here would help the FPGA engineer notice the issue.

maass-hamburg commented on drivers/mdio/Kconfig.xilinx_axi_ethernet_lite:
```diff
@@ -0,0 +1,10 @@
+	bool "Xilinx AXI Ethernet Lite MDIO driver"
+	default y
+	depends on DT_HAS_XLNX_XPS_ETHERNETLITE_3_00_A_MDIO_ENABLED
+	depends on NETWORKING
```
mdio drivers should not depend on NETWORKING

maass-hamburg commented on drivers/mdio/mdio_xilinx_axi_ethernet_lite.c:
```diff
@@ -0,0 +1,233 @@
+
+
+static DEVICE_API(mdio, mdio_xilinx_axi_ethernet_lite_api) = {
+	/* clause 45 not supported */
```
redundant comment, i can see that it is not implemented

WorldofJARcraft replied:
Done

maass-hamburg commented on drivers/mdio/mdio_xilinx_axi_ethernet_lite.c:
```diff
@@ -0,0 +1,233 @@
+
+struct mdio_xilinx_axi_ethernet_lite_data {
+	struct k_mutex mutex;
+	bool bus_enabled;
```
no longer needed as it cant be disabled anymore. Also drivers here don't have to check themself, if they have been correctly init, in their implemented api calls

maass-hamburg commented on drivers/mdio/mdio_xilinx_axi_ethernet_lite.c:
```diff
@@ -0,0 +1,233 @@
+};
+
+static inline uint32_t
+mdio_xilinx_axi_ethernet_lite_read_reg(const struct mdio_xilinx_axi_ethernet_lite_config *config,
```
please shorten these names and also the defines as these are not global or exposed to other files, they don't have to be that unique

maass-hamburg commented on drivers/ethernet/eth_xilinx_axi_ethernet_lite.c:
```diff
@@ -0,0 +1,609 @@
+	data->tx_ping_toggle = !data->tx_ping_toggle;
+
+	/* no need to linearize - can copy fragments one by one into transmit buffer */
+	do {
+		int frag_len = cursor->buf->len;
+		const uint8_t *frag_data = cursor->buf->data;
+
+		xilinx_axi_ethernet_lite_write_transmit_buffer(config, buffer_addr, frag_data,
+							       frag_len);
+
+		buffer_addr += frag_len;
+	} while (xilinx_axi_ethernet_lite_cursor_advance(cursor));
```
as you are just copying the packet at a specific location, just use `net_pkt_read()` directly. in https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/ethernet/eth_litex_liteeth.c it is used similar. You don't have to handle each buffer by its own.

maass-hamburg replied:
Have you tried it?

WorldofJARcraft replied:
I tested it right now, replacing only the code on the RX side:
```C
static inline int axi_eth_lite_read_to_pkt(const struct axi_eth_lite_config *config,
					   struct net_pkt *pkt, mem_addr_t buffer_addr,
					   size_t bytes_to_read)
{
	return net_pkt_write(pkt, (void*)buffer_addr, bytes_to_read);
}
```
The original function was:
```C
static inline int axi_eth_lite_read_to_pkt(const struct axi_eth_lite_config *config,
					   struct net_pkt *pkt, mem_addr_t buffer_addr,
					   size_t bytes_to_read)
{
	int ret = 0;

	for (size_t read_bytes = 0; read_bytes < bytes_to_read; read_bytes += sizeof(uint32_t)) {
		uint32_t current_data = axi_eth_lite_read_reg(config, buffer_addr);
		size_t bytes_to_write_now = read_bytes + sizeof(uint32_t) > bytes_to_read
						    ? bytes_to_read - read_bytes
						    : sizeof(uint32_t);

		ret += net_pkt_write(pkt, &current_data, bytes_to_write_now);

		if (ret < 0) {
			LOG_ERR("Write error bytes %zu/%zu (%zu)", read_bytes, bytes_to_read,
				bytes_to_write_now);
		} else {
			LOG_DBG("Write OK bytes %zu/%zu (%zu) cursor %p remaining %d", read_bytes,
				bytes_to_read, bytes_to_write_now, pkt->cursor.buf,
				pkt->cursor.buf ? pkt->cursor.buf->size - pkt->cursor.buf->len : 0);
		}

		buffer_addr += sizeof(uint32_t);
	}
	return ret;
}
```
The results were interesting: using `net_pkt_write` directly instead of in a loop causes no problems for some packets, but breaks reception for other packets:
<img width="1700" height="673" alt="image" src="https://github.com/user-attachments/assets/39340914-7cc3-43a4-81cd-dabd7cee16d6" />
My assumption is that - as I explained above - packets with a length that is a multiple of `sizeof(long)` like ARP can be copied with no issue, but packets with different lengths cause issues due to byte-wise read and write operations.
Again, the results might be different depending on the implementation of the CPU, bus, minor revisions of the IP, ..., so I would be hesistant to use this approach even if it did work.

maass-hamburg replied:
how about rounding the size up to the next value dividable by 4. Saw something like that in eth_smsc911x.c

maass-hamburg commented on drivers/ethernet/eth_xilinx_axi_ethernet_lite.c:
```diff
@@ -0,0 +1,609 @@
+		buffer_addr += sizeof(uint32_t);
+	}
+
+	hdr = (struct net_eth_hdr *)header_buf;
+	len = hdr->type;
+	/*
+	 * AXI Ethernet Lite cannot tell us the length of the received packet, so we try to parse it
+	 * Also, FCS is not used by Zephyr stack
+	 */
+	switch (ntohs(len)) {
+	case NET_ETH_PTYPE_ARP:
+		/* fixed length */
+		packet_size =
+			sizeof(struct net_eth_hdr) + XILINX_AXI_ETHERNET_LITE_ARP_PACKET_LENGTH;
+		break;
+	case NET_ETH_PTYPE_IP: {
+		const struct net_ipv4_hdr *ip4_hdr =
+			(const struct net_ipv4_hdr *)&header_buf[sizeof(*hdr)];
+		len = ip4_hdr->len;
+		packet_size = ntohs(len);
+		/* length includes ipv4 header length */
+		packet_size += sizeof(struct net_eth_hdr);
+		break;
+	}
+	case NET_ETH_PTYPE_IPV6: {
+		const struct net_ipv6_hdr *ip6_hdr =
+			(const struct net_ipv6_hdr *)&header_buf[sizeof(*hdr)];
+		/* payload + any optional extension headers */
+		len = ip6_hdr->len;
+		packet_size = ntohs(len);
+		packet_size += sizeof(struct net_eth_hdr) + sizeof(*ip6_hdr);
+		break;
+	}
+	default:
+		/* use the full MTU... */
+		break;
+	}
+
```
the len of a ethernet packet is always at the same location, independent of independent of the protocol (ARP, ipv4, ipv6) in it.
use that.

https://docs.amd.com/r/en-US/pg135-axi-ethernetlite/Type/Length

WorldofJARcraft replied:
This type/length field is a part of the Ethernet MAC header.
This only indicates the packet length for the LLC (link-layer control) L2 protocol, a predecessor of modern Ethernet, which is not commonly used any more.
Modern networks generally use this field as a Type indicator (EtherType) instead.
For example, zephyr's network stack always interprets this as a *type* instead of a *length*, which is always the same for each protocol irrespective of the packet length.

For that reason, the IP has a separate *transmit size* MMIO register and does not rely on the type/length field for transmissions.
I have no idea why the same does not exist for RX; this was probably a design oversight.

maass-hamburg commented on drivers/ethernet/eth_xilinx_axi_ethernet_lite.c:
```diff
@@ -0,0 +1,609 @@
+	.set_config = xilinx_axi_ethernet_lite_set_config,
+	.send = xilinx_axi_ethernet_lite_send};
+
+static inline int
+xilinx_axi_ethernet_lite_read_to_pkt(const struct xilinx_axi_ethernet_lite_config *config,
+				     struct net_pkt *pkt, mem_addr_t buffer_addr,
+				     size_t bytes_to_read)
+{
+	int ret = 0;
+
+	for (size_t read_bytes = 0; read_bytes < bytes_to_read; read_bytes += sizeof(uint32_t)) {
+		uint32_t current_data = xilinx_axi_ethernet_lite_read_reg(config, buffer_addr);
+		size_t bytes_to_write_now = read_bytes + sizeof(uint32_t) > bytes_to_read
+						    ? bytes_to_read - read_bytes
+						    : sizeof(uint32_t);
+
+		ret += net_pkt_write(pkt, &current_data, bytes_to_write_now);
+
+		if (ret < 0) {
+			LOG_ERR("Write error bytes %zu/%zu (%zu)", read_bytes, bytes_to_read,
+				bytes_to_write_now);
+		} else {
+			LOG_DBG("Write OK bytes %zu/%zu (%zu) cursor %p remaining %d", read_bytes,
+				bytes_to_read, bytes_to_write_now, pkt->cursor.buf,
+				pkt->cursor.buf ? pkt->cursor.buf->size - pkt->cursor.buf->len : 0);
+		}
+
+		buffer_addr += sizeof(uint32_t);
+	}
+	return ret;
+}
```
just read it at one. take a look at https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/ethernet/eth_litex_liteeth.c for reference

maass-hamburg commented on drivers/ethernet/eth_xilinx_axi_ethernet_lite.c:
```diff
@@ -0,0 +1,609 @@
+	struct k_sem tx_sem;
+	/* used to trigger the RX ISR from time to time */
+	struct k_timer rx_timer;
+	/* used to offload copying an RX packet outside of ISR context */
```
why not copying it in the isr? Would be less overhead

WorldofJARcraft commented on drivers/ethernet/CMakeLists.txt:
```diff
@@ -45,6 +45,7 @@ zephyr_library_sources_ifdef(CONFIG_ETH_LAN9250		eth_lan9250.c)
 zephyr_library_sources_ifdef(CONFIG_ETH_SY1XX		eth_sensry_sy1xx_mac.c)
 zephyr_library_sources_ifdef(CONFIG_ETH_NXP_ENET	eth_nxp_enet.c)
 zephyr_library_sources_ifdef(CONFIG_ETH_XILINX_AXIENET	eth_xilinx_axienet.c)
+zephyr_library_sources_ifdef(CONFIG_XILINX_AXI_ETHERNET_LITE	eth_xilinx_axi_ethernet_lite.c)
```
Done

michalsimek commented:
nit: s/status = "okey";/status = "okay";/

I can't see any reason why new DT binding should be used compare to one which exists for years in Linux/U-Boot. 
I can don't think there is a need to introduce -mac compatible string when you already have xlnx,xps-ethernetlite-3.00.a. And then reference back parent for reg property.
The same is with defining new xlnx,xps-ethernetlite-3.00.a-mdio for mdio. When mdio child is present mdio exists and access can be used. 

mdio is optional but won't work without actual "mac" that's why can't see any reason why it should be described in a way that they work as two separate entities and also two separate drivers. 
From my perspective you should use 

ethernet@xxx {
   compatible = ...
   reg = ...
   mdio {
      ...
   };
};

and create only single driver not two.

WorldofJARcraft commented on drivers/ethernet/Kconfig.xilinx_axi_ethernet_lite:
```diff
@@ -0,0 +1,24 @@
+	default 1
+	depends on ETH_XILINX_AXI_ETHERNET_LITE
+	help
+		Period for periodic RX timer task in AXI Ethernet Lite.
```
Fixed, thanks

WorldofJARcraft commented:
@ibirnbaum thanks for the summary. I would like to add that for the AXI Ethernet Lite (and many other families of Ethernet IP), the MDIO can be configured independent of the MAC or even removed completely at hardware synthesis/elaboration time. For that reason, depending on the device in question, the MDIO might even have separate register ranges or interrupts. For example, Xilinx' AXI Ethernet Subsystem has two interrupts, one for the MDIO and one for optional MAC features, both of which can be disabled at the hardware level.
Hence, to me, it makes sense that the MDIO is a separate device.
Actually, I remember that we had a similar discussion for my other Xilinx ethernet driver, the AXI Ethernet Subsystem (https://github.com/zephyrproject-rtos/zephyr/pull/73986#issuecomment-2650287183) as well. Hence, I oriented myself on how that driver is structured when working on this PR.

I would also like to point out that a more general discussion regarding the MDIO/MAC architecture should probably be had, and a larger refactoring may be necessary.
Discussing this individually for each new driver creates a lot of extra work and has the potential of solving the same problem differently for different devices.

venodela commented on drivers/mdio/mdio_xilinx_axi_ethernet_lite.c:
```diff
@@ -0,0 +1,198 @@
+	return 0;
+}
+
+static int axi_eth_lite_init(const struct device *dev)
```
There is function naming overlap between Axi_ethernet_lite and MDIO drivers (e.g., axi_eth_lite_init, axi_eth_lite_read_reg, axi_eth_lite_write_reg). Please add clear prefixes like mdio_ or unique names in mdio driver.

@michalsimek @HariniKatakamX

venodela commented:
> Here is an exemplary device tree entry for this device:
> 
> ```
> axi-ethernet-lite@40E00000{
> 
> 		/* shared between MAC and MDIO */
> 		reg = <0x40E00000 0x10000>;
> 
>                compatible = "xlnx,xps-ethernetlite-3.00.a";
> 
> 		status = "okay";
> 
> 		arty_ethernet:axi-ethernet-lite-mac {
> 			/* MAC and MDIO cannot be parent and child, this breaks the preprocessing macros */
> 			compatible = "xlnx,xps-ethernetlite-3.00.a-mac";
> 			interrupt-parent = <&plic>;
> 			interrupts = <12 0x4>;
> 			clocks = <&clk_bus_0>;
>       	
> 			local-mac-address = [00 0a 35 00 00 01];
> 			
> 			xlnx,rx-ping-pong;
> 			xlnx,tx-ping-pong;
> 			phy-connection-type = "mii";
> 
> 			status = "okay";
> 		};
> 
> 
> 		axi_ethernetlite_0_mdio: mdio {
> 			#address-cells = <1>;
> 			#size-cells = <0>;
> 
> 			compatible = "xlnx,xps-ethernetlite-3.00.a-mdio";
> 			status = "okay";
> 
> 			arty_phy0: phy@1 {
> 				#address-cells = <1>;
> 				#size-cells = <0>;
> 				compatible = "ethernet-phy";
> 				reg = <1>;
> 				status="okay";
>     			};
> 	  	};
> 	};
> ```


It appears that the PHY node also requires the default-speeds device tree property to restrict link negotiation to 10/100 Mbps, which are the only speeds supported by the AXI Ethernet Lite core.

Without specifying default-speeds, the PHY might advertise 1000 Mbps capability during auto-negotiation, which is not supported by the AXI Ethernet Lite MAC.

References:
https://github.com/zephyrproject-rtos/zephyr/pull/89210/files

https://github.com/zephyrproject-rtos/zephyr/commit/c069fb0961bddfd28529485572007d6f77560654#diff-29ffafb34eab81ce4238d037df9970de5f83200f42bd63cf105307f949e7e26a

@HariniKatakamX

WorldofJARcraft committed:
drivers: ethernet: Add Xilinx Axi Ethernet Lite

This commit adds support for the Xilinx AXI Ethernet Lite device,
also known as the emaclite.
The emaclite is a light-weight 10/100 MII Ethernet device.
It foregoes a DMA to reduce chip area.
Instead, it uses memory-mapped transmit/receive buffers.
A selection of features can optionally be enabled:
- a second ("pong") RX/TX Buffer
- MDIO support
- Interrupt support
This driver handles the MAC functionality of the core;
a driver for the MDIO part is introduced separately
as an MDIO driver.

Signed-off-by: Eric Ackermann <eric.ackermann@cispa.de>

venodela commented:
> Hello @venodela,
> 
> I do not see any review comments from you. Is it possible that they are still in _draft_ state?

@WorldofJARcraft sorry i submitted review comments now;

WorldofJARcraft commented on drivers/ethernet/Kconfig.xilinx_axi_ethernet_lite:
```diff
@@ -0,0 +1,24 @@
+menuconfig ETH_XILINX_AXI_ETHERNET_LITE
+	bool "Xilinx AXI Ethernet Lite Driver"
+	default y
+	depends on DT_HAS_XLNX_XPS_ETHERNETLITE_3_00_A_MAC_ENABLED
```
Done

WorldofJARcraft commented on drivers/mdio/mdio_xilinx_axi_ethernet_lite.c:
```diff
@@ -0,0 +1,198 @@
+	return mdio_control_reg_val & AXI_ETH_LITE_MDIO_CONTROL_REG_MDIO_BUSY_MASK ? -EBUSY : 0;
+}
+
+static inline void axi_eth_lite_set_addr(const struct axi_eth_lite_config *config, uint8_t prtad,
```
Done

venodela commented:
Hi @michalsimek 

My review comments are addressed and this pull request now looks fine to me. Could you please approve.

@WorldofJARcraft @maass-hamburg

pdgendt reviewed (commented):
Mostly nits, looks good overall
